### ISSUE TITLE
Option to disable Media Caching

### ISSUE THREAD

**Issue by kevATin** [tootsuite/mastodon#15195]
### Pitch

Give instance owners the option to disable all caching of external media.
This should include previews of media links that are posted locally by your users, but also media from federated toots.

To be clear, the media links should still work, and ideally even load their preview in the UI, they simply shouldn't ever be downloaded onto the server.

### Motivation

Some server owners might not have the storage space, or time to moderate media.

Together with #15193 this would allow for having an instance that has close to no media whatsoever (both internal and external), severely decreasing the legal responsibilities that come with hosting an instance.

**Comment by Gargron** [comment]
> severely decreasing the legal responsibilities that come with hosting an instance.

I am not a lawyer but I do not believe that your legal responsibilities are decreased if you are "merely linking" to a problematic resource (especially if you are hotlinking like you are asking for here).

Hotlinking comes with a few caveats that make it a bad choice for Mastodon:

- Loading content from 3rd party domains is a privacy issue, as the remote host would get the IP of the users on your server. You would potentially need to make people aware through your privacy policy. Besides being a privacy issue, number of separate DNS lookups on a page would negatively affect client-side loading times
- Potential security issue when serving content that hasn't been put through imagemagick/ffmpeg checks, e.g. image or video files with browser exploits (not aware of any specific exploits like this currently but it's a possibility)
- Can't guarantee the media is in a playable format which leads to frustrating UX. While Mastodon always converts media into web-playable formats on upload, Pleroma for example does no such thing
- No thumbnails, or no thumbnails in the correct dimensions and format, leading to either too much bandwidth usage on both the client and the remote server, or thumbnails that are potentially too blurry for the way they are displayed

**Comment by kevATin** [comment]
Hello @Gargron thank you for your response and sorry for the late reply.

I don't know about the current legal situation, however I do remember there being a lot of talk in European politics regarding whether people who link to something illegal are as liable as those who host it. The fact that it was so hotly debated makes it seem to me like there is value in offering that option.

I can see your points; maybe my idea was too ambitious to put into one feature suggestion, so maybe it would be better to take it step by step.
Disregarding the internal and just focusing on external media: Wouldn't extending the Server-wide Moderation "Reject Media" like mastodon.online does for pawoo.net do pretty much what I am looking for, or am I missing something?

Do you think an instance that treats any other instance as Reject Media could work (by itself and as part of the fediverse), apart from having no thumbnails in the timeline and requiring more clicks from the users?

**Comment by nemobis** [comment]
> whether people who link to something illegal are as liable as those who host it

Copyright-wise in EU, this is mostly about the question of ["making available"](https://duckduckgo.com/?q=%22making+available%22+site%3Aipkitten.blogspot.com&t=ffab&ia=web). The courts are very active in this field. Caching has [specific rules about it](https://eur-lex.europa.eu/legal-content/EN/TXT/HTML/?uri=CELEX:32000L0031&from=en). Do consult a lawyer if you're looking for risk reduction.

**Comment by Arthaey** [comment]
I'd like this option for storage space reasons, regardless of the legal implications.

**Comment by Xipherisroot** [comment]
Yeah, looking for this option pretty bad...

The legal issue is sort of moot if 'merely linking' to a problematic resource is as bad as unwillingly or unwittingly hosting it so that's not much of an argument.

Loading content from a third party should be a choice, its fine and 'good' that the default is leaning towards privacy and security but there's no reason not to give someone running Mastodon a choice, it makes running a private instance far more appealing and cost effective in terms of resource cost.  Franky when it comes to arguments such as DNS lookup time cost and such, that's what DNS caching is for and the web in general works this way already, most websites have content on them pointing to, youtube averages 10 different domain lookups on the home page even with an adblocker installed, same with a site like Tumblr, Techspot, just a small spot check of random sites, this is a non-issue.

As for loading content not put through a processor run by your own server or the server of someone you have an account with, this is something that happens all day long every day for any person casually using the internet, images are loaded from various websites and third parties inline with content all the time, this argument doesn't make a lot of sense? I am certainly not putting every image or video I see while browsing the internet through my own instance of imagemagick or ffmpeg before viewing it...

For formats etc, this is a small price to pay if you enable this option, modern browsers, operating systems and app frameworks have access to fairly broad and decent codec support for playback, someone would at this day and age need to be purposefully obscure in their choices to make it difficult to play back or view media on the platform.

When it comes to the thumbnails, there's elegant and inelegant ways of handling it, everything from have the server generate a thumbnail but show the source image when clicked through to simply showing the source image at the timeline specific scale, this is a choice someone running their instance would actively be making and the limitations inherent is the cost they would need to weight, but the choice should be theirs, providing a choice is never a negative, choices and the right to choose is a positive thing.

I would far prefer the option and ability to disable local caching of content, and frankly as someone running a server wherein I often upload images and videos that the limits of size for the platform to an audience, I am not at all afraid of any silly bandwidth use and neither should larger server operators, hosting content available to the wider web means bearing the cost of providing that content.... to the wider web.... if 500,000 people want to see some 100 mb video I've put up, it'll cost me a couple hundred bucks and that's the cost of providing content, anyone running a system online available to the public must understand this....?

**Comment by mjohanning99** [comment]
Considering the recent problems on other federated services such as Lemmy (lemmy.world), I would also very much like the option of disabling the caching of external media. Is there perhaps a patch I can apply in some .rb file? The only idea I've had so far is running tootctl in the background every x minutes, definitely no ideal.

**Comment by alyxto** [comment]
I never thought I'd say this, but in this case, I agree mostly with [Gargron](https://github.com/mastodon/mastodon/issues/15195#issuecomment-897194803).

**Legal**
The legal aspect varies depending on the region you are in.
But at least for Germany, I can say with confidence that just (hot-) linking to remote content does not free you from any liability.
However, this isn't really an issue as platform operators and hosting providers are usually waived from responsibility for unknown user generated content, considering you have implemented "reasonable" measures and act and remove it within a reasonable amount of time once you will become aware of it (trough for example user reports or other means).

**Security and Privacy**
In regard to security and privacy, embedding remote content is a concern.
Again, we talk about user generated content. So you can't compare this with you embedding an image on your personal blog.
While embedding images and videos is generally speaking considered somewhat safe, it's just been a few months since we had the last vulnerability related to such content (in this case webp).
So processing/validating content from untrusted sources can and should be seen as best practice.

Privacy wise, it's even worse.
Embedding user generated remote content is a total nightmare. It would enable anyone with a bit of technical knowledge to basically "harvest" your user's IPs + some other metadata.
We've talked about this problem decades ago. Forums didn't want to cover the costs of storing the images or feared legal trouble so, often uploads to them directly were not always possible and users were encouraged to upload media to other platforms and just embed them in the forum post.
Then bad actors abused this option with IP grabbers and similar tools, resulting in many Forums to apply whitelists filtering from which sources (mostly image upload services) users are allowed to embed media.

But, due to the federated nature of the Fediverse a whitelist is not really possible. At least not without discriminating smaller / newer instances.

So if you really want to have embedded remote content, and you don't wanna half-ass it, you probably need to do it like email clients do it.
The user needs to have an option to say "Yes, I want to load this image from cdn.example.com".

This would definitely affect the user experience.
A more UX friendly approach would be to use a reverse proxy and proxy the request to the remote content through your mastodon instance.
This is already being done by some fedi software (iirc), and solves some but not all privacy issues. Some tracking methods like read receipts would still work.

Also, it would be kinda inefficient when your server needs to pull the same image multiple times within a short time frame.
Soo, it would make sense to cache the content at least for a somewhat short time frame.
It's also reasonable to assume that the majority (80%?) of the content is consumed within a day or two after your instance become aware of this post in general.
So we could also tackle the tracking issue by caching the remote content directly when we pull the post from the remote instance... 

Which leads us to the system we have right now.

**Media Format**
The "Can't guarantee the media is in a playable format" argument is kinda crappy tho..
The only format I can come up with is JXL, but everyone who is actually touching it is definitely aware that it's rarely supported.
Soo, I guess it would be fine if mastodon would handle such embeds as regular download links, like it is doing already with unknown files from remote instances.
However, I also see a huge advantage for app developers, especially for non-web-based apps, in having a limited and defined list of possible media types.

**UX**
I'm also somewhat active in the Lemmy part of the Fediverse. Caching remote content wasn't/isn't really a well working feature there. So many have disabled it, and some parts were always embedded remote content for a while.

I often experienced it that images and icons were missing or loading slow.
Sometimes remote instances are down, sometimes they are in some far away location and don't use a CDN.
Overall the experience wasn't that great and often affected by remote content. I was often able to tell you exactly which instance is having trouble right now just by looking on my home page/timeline.
It's improved a lot by now, but I assume also because many small, less performant instances died over time.

I have a friend who is into photography. They are on an akkoma instance, and akkoma doesn't seem to really process the uploaded images (by default?).
This is probably good if you like high quality pictures, but at the same time it's an awful experience when you don't have the best internet connection. Sometimes posts contain 6-8 pictures, each 10-15MB,
The pictures are beautiful, but the loading times are awful.. 

On the other side... Mastodon image and video compression is an awful mess too...
It seems to work great for HD images, but fails dramatically at everything else.
Small pictures end up as a big pile of compression artifacts, and all videos I upload so far end up being bigger than what they were before.. (same file format)
So, I guess remote content would fix this issue xD


**Shared Costs & Responsibility in a federated social network**
I personally think that caching remote content is, at least partially, a social responsibility for instance admins.
Most fedi instances (especially smaller once) operate at a loss, so even after potential donations, the admin is still paying a significant portion of the bills.
Most use some S3 storage and many even use CDNs to offer a good performance to their users.
Both are usually paid per used GB (+ sometimes per req), or in other words.. every time you download something from their instance, you increase their monthly bill.
Sure, its a tiny tiny amount, but it still adds up to a not so tiny number as many admins probably can confirm.

Caching remote content reduces the burden every instance has to bear, the same way we share moderation efforts.


**tl;dr**
I don't see that many valid reasons that out wight the negative effects.
<ins>If you're having issues regarding the required storage, lower the "Media Cache Retention Period".</ins>

But, I also don't like when others made decisions for me. Which.. Eugen likes to do 😅
So, I am still in favor of giving admins the option, **but** if this will become an option, everyone should be aware of the risks and issues that comes with it. Especially the **users**.
Therefore, I would be for an "email client" like approach where users have to explicitly confirm from which resource/server they want to load content.

**Comment by evrial** [comment]
> But, I also don't like when others made decisions for me.
this should be tldr.

**Comment by alyxto** [comment]
@evrial I guess is a whole different discussion if mastodon is the project that likes to give users options.
We had that multiple times, and I guess by now we are aware of the position they're having.

When you run a single user instance, you're probably know what you are doing. Even if not, then you are at least aware that you don't know what you are doing 😅 

But as soon as you have some other users on your instance you become the one who makes decisions for them.
I know there is often no way around it, but I would like if users are at least aware of it and the risks that come with it. Especially when it comes to privacy.

I've seen admins disability the cache feature with the intent to save a few bucks without thinking too much abut it.
And I've seen someone wondering how a random creepy dude figured out where they live just by sending them a DM.

People often see their instance as somewhat trustworthy, at least more trustworthy than some random / shady website.
And I think this is a fair assumption. I mean.. I would not share my SSN with my instance admin, but I trust them enough that I don't open the Tor browser or a VPN to access the instance.

When just everybody can embed content to your feed and messages, without you even knowing or understanding it, it becomes a security/privacy concern.

This is why I am in favor of an option where users have to confirm it.

**Comment by kevATin** [comment]
> Therefore, I would be for an "email client" like approach where users have to explicitly confirm from which resource/server they want to load content.

I like that as it would give both instance owners as well as users more options. 
I'm well aware that for a big percentage it won't make sense to do it this way, but i think admins and users should still have the agency over whether they want it that way or not.

> Most use some S3 storage and many even use CDNs to offer a good performance to their users. 

It might be the case that most instances use big hosting providers, however i don't think it's a good idea to plan Mastodon development based on that. To me Mastodon is free because you can self-host it at home without depending on anyone. If the default setup assumes that I am just installing Mastodon on a rented server in a Amazon, Microsoft, Google, IBM, etc datacenter and is optimized for exactly this use case then it's not very free nor independent. 

> email client

I could see it working like this: If the admin decides to disable all local media caching then users are warned of this during registration and have to manually approve individual domains for loading previews (including the option to make a wildcard entry followed by a big red warning).
And this could also give users of instances that have these media previews enabled more options. If the admin has it enabled the user could get the option to either use the offered cached previews or decide to get them directly from the source-instance.

Further up I had mentioned the "reject media" moderation option that was added for pawoo and similar instances.
I'm not sure how that is currently handled, but didn't you have to specifically visit that instance to view the images from those reject media instances? 
If both instances and users had more options when it comes to who loads media how, then this might be a better solution to those kinds of issues as well.
So in the future for these kinds of reject media posts your instance could inform you that they do not directly cache this image and you get the option of moving on or fetching it directly from the source. In case this is already more or less how it's done in this case then i'd say it makes even more sense to give users and admins this option for everything else too.

**Comment by evrial** [comment]
@alyxto I don't any reason to pull the security card because attack surface of Mastodon is large as Jupiter. Simply use latest browser and VPN and Mastodon server should only inform when you create an account about all that condition.

**Comment by hypermug1** [comment]
Friendica, Pleroma, and Misskey are all popular softwares and proxy media rather than storing all of it. I see no need for mastodon to do this. Just necroposting dont mind me.

**Comment by tesaguri** [comment]
> So if you really want to have embedded remote content, and you don't wanna half-ass it, you probably need to do it like email clients do it.
> The user needs to have an option to say "Yes, I want to load this image from cdn.example.com".
> 
> This would definitely affect the user experience. […]

Mastodon has already had an <q>Always hide media</q> setting, and I think the UX of the <q>email client</q> approach would no different from the existing feature.

**Comment by evrial** [comment]
Ok unsubbing, I don't see any interest waiting and mastodon is a piece of garbage.

**Comment by alyxto** [comment]
@evrial You are free to go. Nobody ever forced you to use mastodon in the first place.
Or to share your rubbish option that we are all servants of the capitalist class.

**Comment by alyxto** [comment]
> Mastodon has already had an "Always hide media" setting, and I think the UX of the "email client" approach would no different from the existing feature.

Always hide is not comparable to a "load this resource from x" in my opinion.

It's not about enabling or disabling media.
It's about giving the user the option to knowingly load media and share personal information with "totalllytrustworthy.ru"


But again, for me **personally**, it's questionable why we load remote content in the first place.
I think a proxy solution should be the bare minimum, and in this scenario, the question would be: what issues are causing the current limitation that prevents reducing the proxy cache to less than 1 day?

**Comment by aakoskin** [comment]
For multimedia content there is a simple kludge available by redefining FFMPEG_BINARY/FFPROBE_BINARY to something else in .env.production (e.g. /bin/false). This nicely avoids storing of files, but unfortunately Mastodon apparently ignores the return value and still presents a playback window which of course fails. IMHO, this is a bug; it should give a "preview not available" (like it sometimes does) and link to the original post/site.

I'm just posting this tip here in case someone is struggling with media content overhead on a small server. This helps also with ffmpeg resource usage getting out of hand and chocking the server (https://github.com/mastodon/mastodon/issues/16496).

**Comment by huser123** [comment]
Yes please!
